plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java

```diff
@@ -20,61 +20,127 @@
+        Duration caseInsensitiveNameMatchingCacheTtl = requireNonNull(config.getCaseInsensitiveNameMatchingCacheTtl(), "caseInsensitiveNameMatchingCacheTtl is null");
+
+        this.caseInsensitiveNameMatching = config.isCaseInsensitiveNameMatching();
+        CacheBuilder<Object, Object> remoteNamesCacheBuilder = CacheBuilder.newBuilder()
+                .expireAfterWrite(caseInsensitiveNameMatchingCacheTtl.toMillis(), MILLISECONDS);
+        this.remoteDatasetNames = remoteNamesCacheBuilder.build();
+        this.remoteTableNames = remoteNamesCacheBuilder.build();
+    }
+
+    String toRemoteDatasetName(String projectId, String datasetName)
+    {
+        requireNonNull(projectId, "projectId is null");
+        requireNonNull(datasetName, "datasetName is null");
+        verify(CharMatcher.forPredicate(Character::isUpperCase).matchesNoneOf(datasetName), "Expected schema name from internal metadata to be lowercase: %s", datasetName);
+        if (caseInsensitiveNameMatching) {
+            try {
+                @Nullable String remoteDatasetName = remoteDatasetNames.getIfPresent(datasetName);
+                if (remoteDatasetName == null) {
+                    // this might have been just created, reload cache
+                    Map<String, String> mapping = stream(listDatasets(projectId))
+                            // will throw on collisions to avoid ambiguity
+                            .collect(toImmutableMap(dataset -> dataset.getDatasetId().getDataset().toLowerCase(ENGLISH), dataset -> dataset.getDatasetId().getDataset()));
+                    remoteDatasetNames.putAll(mapping);
+                }
+                remoteDatasetName = remoteDatasetNames.getIfPresent(datasetName);
+                if (remoteDatasetName != null) {
+                    return remoteDatasetName;
+                }
+            }
+            catch (RuntimeException e) {
+                throw new TrinoException(BIGQUERY_OBJECT_NOT_FOUND, "Failed to find remote dataset name: " + firstNonNull(e.getMessage(), e), e);
+            }
+        }
+
+        return datasetName;
     }
 
-    TableInfo getTable(TableId tableId)
+    String toRemoteTableName(String projectId, String remoteDatasetName, String tableName)
     {
-        TableId bigQueryTableId = tableIds.get(tableId);
-        Table table = bigQuery.getTable(bigQueryTableId != null ? bigQueryTableId : tableId);
-        if (table != null) {
-            tableIds.putIfAbsent(tableId, table.getTableId());
-            datasetIds.putIfAbsent(toDatasetId(tableId), toDatasetId(table.getTableId()));
+        requireNonNull(projectId, "projectId is null");
+        requireNonNull(remoteDatasetName, "remoteDatasetName is null");
+        requireNonNull(tableName, "tableName is null");
+        verify(CharMatcher.forPredicate(Character::isUpperCase).matchesNoneOf(tableName), "Expected table name from internal metadata to be lowercase: %s", tableName);
```

Comment:
how does this one work for non-letter characters?

Reply:
Good catch. I think I should use https://docs.oracle.com/javase/8/docs/api/java/lang/Character.html#isUpperCase-int-

I'll fix and add tests to cover these cases.

Reply:
I wish Java would let go of UTF-16 already... for anyone who come along in the future the correct way to check this is like

```suggestion
        verify(tableName.codePoints().noneMatch(Character::isUpperCase), "Expected table name from internal metadata to be lowercase: %s", tableName);
```

Reply:
To be honest I rather meant the non-letter characters like hyphens, underscosers etc. But nice you though of fancy UTF stuff. Can those be used in names in BQ?